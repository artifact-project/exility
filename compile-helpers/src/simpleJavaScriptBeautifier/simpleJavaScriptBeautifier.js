"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tabs = '\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t';
var R_IF_ELSE = /^}\s*(if|else)/;
var R_RETURN = /return.*?;$/;
var R_BLOCK_CLOSE = /^}[;,]?$/;
var R_BLOCK_OPEN = /(function(\s\w+)?|if|for|return(\sfunc.*?)?)\s*[\(\{]/;
var R_VAR_OPEN = /^var.+(\{|\(|,)$/;
var R_INVOKE_BLOCK_OPEN = /^[a-z_][a-z0-9_]*\([^)]*$/i;
var R_INVOKE_BLOCK_CLOSE = /^\}?\);?$/i;
var R_EMPTY_LINE = /^\s+$/;
function combine() {
    var rules = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        rules[_i] = arguments[_i];
    }
    return new RegExp(rules.map(function (rule) { return rule.source; }).join('|'));
}
var R_OPEN = combine(R_BLOCK_OPEN, R_INVOKE_BLOCK_OPEN, R_VAR_OPEN);
var R_CLOSE = combine(R_BLOCK_CLOSE, R_INVOKE_BLOCK_CLOSE);
function simpleJavaScriptBeautifier(source) {
    var indent = 0;
    return source
        .trim()
        .split('\n')
        .map(function (line) {
        line = line.trim();
        if (R_IF_ELSE.test(line)) {
            line = tabs.substr(0, indent - 1) + line;
        }
        else if (R_OPEN.test(line)) {
            line = tabs.substr(0, indent) + line;
            !R_RETURN.test(line) && indent++;
        }
        else {
            R_CLOSE.test(line) && indent--;
            line = tabs.substr(0, indent) + line;
        }
        return R_EMPTY_LINE.test(line) ? '' : line;
    })
        .join('\n')
        .replace(/\n{2,}/g, '\n\n')
        .replace(/\{\n+/g, '{\n')
        .replace(/\n+(\s+\})/g, '\n$1');
}
exports.default = simpleJavaScriptBeautifier;
