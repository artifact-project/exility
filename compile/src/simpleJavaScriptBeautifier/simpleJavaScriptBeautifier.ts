const tabs = '\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t';

const R_IF_ELSE = /^}\s*(if|else)/;
const R_RETURN = /return.*?;$/;
const R_BLOCK_CLOSE = /^}[;,]?$/;
const R_BLOCK_OPEN = /(function(\s\w+)?|if|for|return(\sfunc.*?)?)\s*[\(\{]/;
const R_VAR_OPEN = /^var.+(\{|\(|[^\}],)$/;
const R_INVOKE_BLOCK_OPEN = /^[a-z_][a-z0-9_]*\([^)]*$/i;
const R_INVOKE_BLOCK_CLOSE = /^\}?\);?$/i;
const R_EMPTY_LINE = /^\s+$/;

function combine(...rules: RegExp[]): RegExp {
	return new RegExp(rules.map(rule => rule.source).join('|'));
}

const R_OPEN = combine(
	R_BLOCK_OPEN,
	R_INVOKE_BLOCK_OPEN,
	R_VAR_OPEN,
);
const R_CLOSE = combine(
	R_BLOCK_CLOSE,
	R_INVOKE_BLOCK_CLOSE,
);

function simpleJavaScriptBeautifier(source: string) {
	let indent = 0;

	return source
		.trim()
		.split('\n')
		.map((line) => {
			line = line.trim();

			if (R_IF_ELSE.test(line)) {
				line = tabs.substr(0, indent - 1) + line;
			} else if (R_OPEN.test(line)) {
				line = tabs.substr(0, indent) + line;
				!R_RETURN.test(line) && indent++;
			} else {
				R_CLOSE.test(line) && indent--;
				line = tabs.substr(0, indent) + line;
			}

			return R_EMPTY_LINE.test(line) ? '' : line;
		})
		.join('\n')
		.replace(/\n{2,}/g, '\n\n')
		.replace(/\{\n+/g, '{\n')
		.replace(/\n+(\s+\})/g, '\n$1')
	;
}

export default simpleJavaScriptBeautifier;
