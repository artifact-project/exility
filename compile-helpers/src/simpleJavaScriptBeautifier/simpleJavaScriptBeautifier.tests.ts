import simpleJavaScriptBeautifier
from './simpleJavaScriptBeautifier';

it('empty', () => {
	expect(simpleJavaScriptBeautifier('')).toBe('');
});

it('function', () => {
	expect(simpleJavaScriptBeautifier(`
		function foo() {
			var a = 1;
						var b = 2;
		var c = 3;
		return a + b + c; 
			}
	`)).toMatchSnapshot();
});

it('function > function', () => {
	expect(simpleJavaScriptBeautifier(`
		function foo() {
		function bar() {
		}
		return bar;
			}
	`)).toMatchSnapshot();
});

it('function > return function', () => {
	expect(simpleJavaScriptBeautifier(`
function templateFactory() {
"use strict"
return function compiledTemplate(__SCOPE__) {
__SCOPE__ = __SCOPE__ || {};
// CODE:START
var __ROOT = "<!DOCTYPE html>";
return __ROOT
// CODE:END
}
}
`)).toMatchSnapshot();
});

it('if', () => {
	expect(simpleJavaScriptBeautifier(`
		if (expr) {
	expr = false;
		}
	`)).toMatchSnapshot();
});

it('if/else', () => {
	expect(simpleJavaScriptBeautifier(`
		if (expr) {
	expr = false;
				} else {
expr = true;
		}
	`)).toMatchSnapshot();
});

it('if/else if', () => {
	expect(simpleJavaScriptBeautifier(`
if (expr) {
			expr = false;
					} else if (!expr) {
expr = true;
		}
	`)).toMatchSnapshot();
});

it('if/else if/else', () => {
	expect(simpleJavaScriptBeautifier(`
if (expr) {
			expr = false;
					} else if (!expr) {
expr = true;
		} else {
	// LOL!
		}
	`)).toMatchSnapshot();
});


it('invoke', () => {
	expect(simpleJavaScriptBeautifier(`
		call(
		a,
	b,
			c
);
	`)).toMatchSnapshot();
});

it('var object', () => {
	expect(simpleJavaScriptBeautifier(`
function () {
				var __ctx = __STDDOM_CONTEXT();
		var __super__ = {
		"children": __PE_children_2
	};
		var __frag = __STDDOM_FRAGMENT();
}
	`)).toMatchSnapshot();
});

it('call multiline', () => {
	expect(simpleJavaScriptBeautifier(`
__CMP_ATTRS["name"] = (x);
		var __x1 = __STDDOM_CMP_CREATE_INLINE(__CMP_INLINE_STORE,

		__frag,
			__this__,
			"Icon",
		__CMP_ATTRS,
	null
	);
	`)).toMatchSnapshot();
});
