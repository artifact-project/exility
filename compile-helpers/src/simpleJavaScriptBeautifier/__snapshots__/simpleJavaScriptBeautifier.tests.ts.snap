// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call multiline 1`] = `
"__CMP_ATTRS[\\"name\\"] = (x);
var __x1 = __STDDOM_CMP_CREATE_INLINE(__CMP_INLINE_STORE,

	__frag,
	__this__,
	\\"Icon\\",
	__CMP_ATTRS,
	null
);"
`;

exports[`function > function 1`] = `
"function foo() {
	function bar() {
	}
	return bar;
}"
`;

exports[`function > return function 1`] = `
"function templateFactory() {
	\\"use strict\\"
	return function compiledTemplate(__SCOPE__) {
		__SCOPE__ = __SCOPE__ || {};
		// CODE:START
		var __ROOT = \\"<!DOCTYPE html>\\";
		return __ROOT
		// CODE:END
	}
}"
`;

exports[`function 1`] = `
"function foo() {
	var a = 1;
	var b = 2;
	var c = 3;
	return a + b + c;
}"
`;

exports[`if 1`] = `
"if (expr) {
	expr = false;
}"
`;

exports[`if/else 1`] = `
"if (expr) {
	expr = false;
} else {
	expr = true;
}"
`;

exports[`if/else if 1`] = `
"if (expr) {
	expr = false;
} else if (!expr) {
	expr = true;
}"
`;

exports[`if/else if/else 1`] = `
"if (expr) {
	expr = false;
} else if (!expr) {
	expr = true;
} else {
	// LOL!
}"
`;

exports[`invoke 1`] = `
"call(
	a,
	b,
	c
);"
`;

exports[`panel 1`] = `
"function templateFactory(__DEPS__) {
	function panel(attrs, __slots) {
		var __super = {attrs: attrs},
		__default = __slots && __slots.__default || __super.__default
		var __ROOT = \\"<h1>\\" + __STDLIB_HTML_ENCODE(title) + \\"</h1><p>\\";
		return __ROOT
	}

	return function compiledTemplate(__SCOPE__) {
		var __ROOT = panel({title: \\"?!\\"}) + panel({title: \\"Wow!\\"}, {__default: function __default() {
			return \\"Done\\";
			}});

			return __ROOT
		}
	}"
`;

exports[`var object 1`] = `
"function () {
	var __ctx = __STDDOM_CONTEXT();
	var __super__ = {
		\\"children\\": __PE_children_2
	};
	var __frag = __STDDOM_FRAGMENT();
}"
`;
